Python bindings don't compile

```
error[E0061]: this function takes 3 parameters but 2 parameters were supplied
   --> src/lib.rs:242:1
    |
242 | #[py::methods]
    | ^^^^^^^^^^^^^^ expected 3 parameters

error[E0308]: mismatched types
   --> src/lib.rs:242:1
    |
242 | #[py::methods]
    | ^^^^^^^^^^^^^^ expected struct `portus::lang::Scope`, found ()
    |
    = note: expected type `portus::lang::Scope`
               found type `()`

error: aborting due to 2 previous errors

Some errors occurred: E0061, E0308.
For more information about an error, try `rustc --explain E0061`.
error: Could not compile `pyportus`.
```

Bindings probably have to be updated to reflect https://github.com/ccp-project/portus/pull/38

--- a/.travis.yml
+++ b/.travis.yml
@@ -2,5 +2,8 @@ language: rust
 rust:
     - nightly
 cache: cargo
+install:
+  - pip install --user setuptools
+  - pip install --user setuptools_rust
 script:
     - make travis

--- a/Makefile
+++ b/Makefile
@@ -1,6 +1,6 @@
 all: build test-portus test-ipc libccp-integration lint algs
 
-travis: build test-portus libccp-integration
+travis: build test-portus libccp-integration bindings
 
 OS := $(shell uname)
 CLIPPY := $(shell cargo --list | grep -c clippy)
@@ -61,3 +61,10 @@ endif
 ifeq ($(OS), Darwin)
 	cd integration_tests/libccp_integration && export DYLD_LIBRARY_PATH=./libccp && cargo run --bin test_ccp
 endif
+
+.PHONY: bindings python
+
+bindings: python
+
+python:
+	$(MAKE) -C python/

--- a/python/Cargo.toml
+++ b/python/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "pyportus"
-version = "0.2.0"
+version = "0.3.3"
 authors = ["Frank Cangialosi <frankc@csail.mit.edu>"]
 
 [dependencies]

--- a/python/Makefile
+++ b/python/Makefile
@@ -1,0 +1,5 @@
+.PHONY: test-build
+test-build: src/lib.rs
+	mkdir -p build
+	PYTHONPATH='build/' python setup.py develop --install-dir build/
+	rm -rf build/

--- a/python/aimd.py
+++ b/python/aimd.py
@@ -4,14 +4,9 @@ import portus
 class AIMD(portus.AlgBase):
     INIT_CWND = 10
 
-    def on_create(self):
-        self.init_cwnd = float(self.datapath_info.mss * AIMD.INIT_CWND)
-        self.cwnd = self.init_cwnd
-        self.install_program()
-
-    def install_program(self):
-        self.datapath.install(
-	    """\
+    def init_programs(self):
+        return [
+            ("default", """\
                 (def (Report
                     (volatile acked 0) 
                     (volatile sacked 0) 
@@ -37,11 +32,16 @@ class AIMD(portus.AlgBase):
                     (report)
                     (:= Micros 0)
                 )
-            """.format(self.cwnd)
-	)
+            """),
+        ]
+
+    def on_create(self):
+        self.init_cwnd = float(self.datapath_info.mss * AIMD.INIT_CWND)
+        self.cwnd = self.init_cwnd
+        self.datapath.set_program("default", [("Cwnd", self.cwnd)])
 
     def handle_timeout(self):
-        sys.stdout.write("Timeout! Ignoring...\n")
+        sys.stderr.write("Timeout! Ignoring...\n")
 
     def on_report(self, r):
         if r.timeout:
@@ -55,7 +55,6 @@ class AIMD(portus.AlgBase):
 
         self.cwnd = max(self.cwnd, self.init_cwnd)
         self.datapath.update_field("Cwnd", int(self.cwnd))
-
 
 
 portus.connect("netlink", AIMD, debug=True, blocking=True)

--- a/python/portus/__init__.py
+++ b/python/portus/__init__.py
@@ -7,6 +7,7 @@ from . import util, checker
 
 ### Class ### 
 method_signatures = {
+    'init_programs' : ['self'],
     'on_create' : ['self'],
     'on_report' : ['self', 'r']
 }

--- a/python/portus/checker.py
+++ b/python/portus/checker.py
@@ -13,23 +13,28 @@ class ProgramFinder(ast.NodeVisitor):
         self.src_file_name = src_file_name
         self.progs = []
 
-    def visit_Call(self, call_node):
-        for child in ast.walk(call_node):
-            if isinstance(child, _ast.Attribute) and child.attr == "install":
-                args = call_node.args
-                if len(args) > 2:
-                    raise ValueError("datapath.install expects a datapath program string and (optionally) a list of tuples of fields to set at the same time")
-                arg = args[0]
-                found_string_arg = False
-                for arg_child in ast.walk(arg):
-                    if isinstance(arg_child, _ast.Str):
-                        self.progs.append(DatapathProgram(arg_child.s, child.lineno))
-                        found_string_arg = True
-                if not found_string_arg:
-                    raise ValueError("datapath.install must be passed a datapath program as a string literal (not a variable defined elsewhere)")
+    def visit_FunctionDef(self, fd_node):
+        if fd_node.name == "init_programs":
+            for elem in fd_node.body:
+                if isinstance(elem, _ast.Return):
+                    ret_node = elem
+                    if not isinstance(ret_node.value, _ast.List):
+                        raise ValueError("init_programs() must return a list")
+                    for prog in ret_node.value.elts:
+                        if not isinstance(prog, _ast.Tuple):
+                            raise ValueError("init_programs() must return a list of *tuples*")
+                        args = prog.elts
+                        if not isinstance(args[1], _ast.Str):
+                            raise ValueError("init_programs() must return a list of tuples of (2) *strings* (must be string literal)")
+                        self.progs.append(DatapathProgram(args[1].s, prog.lineno))
 
 def _find_datapath_programs(cls):
     src_file_name = inspect.getfile(cls)
+    # NOTE: if module is imported, getfile will return the binary (pyc) rather than the source
+    # This is a hack that assumes the source file is in the exact same directory
+    # (i.e. getfile returns /path/to/x.pyc and we hope the source code is in /path/to/x.py)
+    if '.pyc' in src_file_name:
+        src_file_name = src_file_name[:-1]
     f = open(src_file_name)
     src = ''.join(f.readlines())
     f.close()
@@ -48,8 +53,11 @@ class Colors:
     END = '\033[0m'
     BOLD = '\033[1m'
     BOLDRED = '\033[91;1m'
+    BOLDYELLOW = '\033[93;1m'
     UNDERLINE = '\033[4m'
 
+def bold_yellow_text(t):
+    return Colors.BOLDYELLOW + t + Colors.END
 def bold_red_text(t):
     return Colors.BOLDRED + t + Colors.END
 def bold_text(t):
@@ -62,10 +70,10 @@ def _check_datapath_programs(cls):
         ret = _try_compile(prog.code)
         if ret != "":
             any_errors = True
-            sys.stderr.write("Traceback (datapath program compile error):\n  File \"{}\", line {}\n{}\n{}: {}\n".format(
+            sys.stderr.write("Traceback (datapath program compile error):\n  File \"{}\", {}\n{}\n{}: {}\n".format(
                 pf.src_file_name,
-                prog.lineno,
-                prog.code,
+                bold_yellow_text("line " + str(prog.lineno)),
+                "|\n|" + "\n|".join(prog.code.split("\n")),
                 bold_red_text("error"),
                 bold_text(ret)
             ))

--- a/python/setup.py
+++ b/python/setup.py
@@ -3,7 +3,7 @@ from setuptools_rust import Binding, RustExtension
 
 setup(
     name='portus',
-    version='0.2',
+    version='0.3.3',
     rust_extensions=[RustExtension(
         'portus.pyportus',
         'Cargo.toml',

--- a/python/src/lib.rs
+++ b/python/src/lib.rs
@@ -1,4 +1,4 @@
-#![feature(box_patterns, proc_macro, specialization, const_fn, proc_macro_path_invoc)]
+#![feature(box_patterns, specialization, const_fn)]
 use std::rc::{Rc,Weak};
 
 #[macro_use]
@@ -85,6 +85,60 @@ impl<T: Ipc> CongAlg<T> for PyAlg {
     fn name() -> String {
         // TODO if we want the actual name, need to re-define name() to take &self
         String::from("python")
+    }
+
+    fn init_programs(cfg: Config<T, PyAlg>) -> Vec<(String, String)> {
+        let py = cfg.config.py;
+
+        let py_alg_inst =
+            py_create_flow(py, cfg.config.alg_class).unwrap_or_else(|e| {
+                e.print(py); panic!("Failed to instantiate python class")
+            });
+
+        let programs = match py_alg_inst.call_method0(py, "init_programs") {
+            Ok(ret) => {
+                let list : &PyList = match ret.extract(py) {
+                    Ok(l) => l,
+                    Err(e) => {
+                        e.print(py);
+                        panic!("init_programs() must return a *list* of tuples of (2) strings.\nreturn value was not a list.")
+                    }
+                };
+                let programs : Vec<(String, String)> = list.iter().map(|tuple_obj| {
+                    let tuple : &PyTuple = match tuple_obj.extract() {
+                        Ok(t) => t,
+                        Err(e) => {
+                            e.print(py);
+                            panic!("init_programs() must return a list of *tuples* of (2) strings.\ngot a list, but the elements were not tuples.")
+                        }
+                    };
+                    if tuple.len() != 2 {
+                        panic!(format!("init_programs() must return a list of tuples of *(2)* strings.\ngot a list of tuples, but a tuple had {} elements, not 2.", tuple.len()));
+                    }
+                    let program_name = match PyString::try_from(tuple.get_item(0)) {
+                        Ok(pn) => pn.to_string_lossy().into_owned(),
+                        Err(_) => {
+                            panic!("init_programs() must return a list of tuples of (2) *strings*.\ngot a list of tuples, but the first element was not a string.")
+                        }
+                    };
+                    let program_string = match PyString::try_from(tuple.get_item(1)) {
+                        Ok(ps) => ps.to_string_lossy().into_owned(),
+                        Err(_) => {
+                            panic!("init_programs() must return a list of tuples of (2) *strings*.\ngot a list of tuples, but the second element was not a string.")
+                        }
+                    };
+                    (program_name, program_string)
+                }).collect::<Vec<(String, String)>>();
+                programs
+            }
+            Err(e) => {
+                e.print(py);
+                panic!("error calling init_programs()");
+            }
+        };
+
+        // TODO figure out how to deallocate and clean up instance
+        programs
     }
 
     fn create(control:Datapath<T>, cfg:Config<T, PyAlg>, info:portus::DatapathInfo) -> Self {
@@ -217,6 +271,7 @@ impl<T: Ipc> CongAlg<T> for PyAlg {
             }
         };
     }
+    // TODO implement close, deallocate memory from class
 }
 
 #[py::proto]
@@ -281,7 +336,7 @@ impl PyDatapath {
         };
 
         let ret = {
-            let items : Result<Vec<(String,u32)>, _> = fields.into_iter().map(|tuple_ref| {
+            let items : Vec<(String,u32)> = fields.into_iter().map(|tuple_ref| {
                 let tuple_obj : PyObject = tuple_ref.into();
                 let tuple:&PyTuple = match tuple_obj.extract(py) {
                     Ok(t) => t,
@@ -301,71 +356,68 @@ impl PyDatapath {
                     Err(_) => raise!(TypeError, "second argument to datapath.update_fields must be a list of tuples of the form (string, int|float)")
                 };
                 Ok((name,val))
-            }).collect();
-            items.map(|v| {
-                let args: Vec<(&str,u32)> = v.iter().map(|(s,i)| (&s[..],i.clone())).collect();
-                self.backend.update_field(sc, &args[..])
-            })
+            }).collect::<Result<Vec<(String, u32)>, _>>().unwrap();
+
+            let args: Vec<(&str,u32)> = items.iter().map(|(s,i)| (&s[..],i.clone())).collect();
+            self.backend.update_field(sc, &args[..])
         };
 
         match ret {
-            Ok(Ok(())) => Ok(()),
-            Ok(Err(e)) => { raise!(Exception, format!("Failed to update fields, err: {:?}", e)) },
-            Err(e)     => { raise!(Exception, format!("Failed to update fields, err: {:?}", e)) },
+            Ok(()) => Ok(()),
+            Err(e) => { raise!(Exception, format!("Failed to update fields, err: {:?}", e)) },
         }
         
     }
 
-    fn install(&mut self, py : Python, prog: String, fields : Option<&PyList>) -> PyResult<()> {
+    fn set_program(&mut self, py : Python, program_name: String, fields : Option<&PyList>) -> PyResult<()> {
         if self.debug {
             self.logger.as_ref().map(|log| {
-                debug!(log, "Installing new datapath program";
+                debug!(log, "switching datapath programs";
                    "sid" => self.sock_id,
+                   "program_name" => program_name.clone(),
                )
             });
         }
 
-        let ret = match fields {
+        let ret : Result<Scope, _> = match fields {
             Some(list) => {
-                let items : Result<Vec<(String,u32)>, _> = list.into_iter().map(|tuple_ref| {
+                let items : Vec<(String,u32)> = list.into_iter().map(|tuple_ref| {
                     let tuple_obj : PyObject = tuple_ref.into();
                     let tuple:&PyTuple = match tuple_obj.extract(py) {
                         Ok(t) => t,
                         Err(_) => {
-                            raise!(TypeError, "second argument to datapath.install must be a list of tuples")
+                            raise!(TypeError, "second argument to datapath.set_program must be a list of tuples")
                         }
                     };
                     if tuple.len() != 2 {
-                        raise!(TypeError, "second argument to datapath.install must be a list of tuples with exactly two values each");
+                        raise!(TypeError, "second argument to datapath.set_program must be a list of tuples with exactly two values each");
                     }
                     let name = match PyString::try_from(tuple.get_item(0)) {
                         Ok(ps) => ps.to_string_lossy().into_owned(),
-                        Err(_) => raise!(TypeError, "second argument to datapath.install must be a list of tuples of the form (string, int|float)"),
+                        Err(_) => raise!(TypeError, "second argument to datapath.set_program must be a list of tuples of the form (string, int|float)"),
                     };
                     let val = match tuple.get_item(1).extract::<u32>() {
                         Ok(v) => v,
-                        Err(_) => raise!(TypeError, "second argument to datapath.install must be a list of tuples of the form (string, int|float)")
+                        Err(_) => raise!(TypeError, "second argument to datapath.set_program must be a list of tuples of the form (string, int|float)")
                     };
                     Ok((name,val))
-                }).collect();
-                items.map(|v| {
-                    let args: Vec<(&str,u32)> = v.iter().map(|(s,i)| (&s[..],i.clone())).collect();
-                    self.backend.install(prog.as_bytes(), Some(&args[..]))
-                })
-            },
-            None => { Ok(self.backend.install(prog.as_bytes(), None)) }
+                }).collect::<Result<Vec<(String, u32)>, _>>().unwrap();
+
+                let args: Vec<(&str, u32)> = items.iter().map(|(s,i)| (&s[..],i.clone())).collect();
+                self.backend.set_program(program_name, Some(&args[..]))
+            }
+            None => {
+                self.backend.set_program(program_name, None)
+            }
         };
 
         match ret {
-            Ok(Ok(sc)) => {
+            Ok(sc) => {
                 self.sc = Some(Rc::new(sc));
                 Ok(())
             }
-            Ok(Err(e)) => {
-                raise!(Exception, format!("Failed to install datapath program: {:?}", e))
-            }
             Err(e) => {
-                raise!(Exception, format!("Failed to install datapath program: {:?}", e))
+                raise!(Exception, format!("Failed to set datapath program: {:?}", e))
             }
         }
     }
@@ -391,6 +443,7 @@ fn init_mod(py:pyo3::Python<'static>, m:&PyModule) -> PyResult<()> {
 
     Ok(())
 }
+
 
 use portus::lang;
 use std::error::Error;

